src/Book.ts

```diff
@@ -322,21 +343,82 @@ export class Book<U extends ITransaction = ITransaction, J extends IJournal = IJ
+    // https://www.mongodb.com/docs/v6.0/core/transactions/#read-concern-write-concern-read-preference
+    // Otherwise, we are free to use any readPreference.
+    if (options && !options.session && !options.readPreference) {
+      // Let's try reading from the secondary node, if available.
+      options.readPreference = "secondaryPreferred";
+    }
+
+    const query = { book: this.name } as IFilter;
+
+    let listAccountsSnapshot: IBalance | null = null;
+    if (this.balanceSnapshotSec) {
+      listAccountsSnapshot = await getBestListAccountsSnapshot({ book: this.name });
+      if (listAccountsSnapshot) {
+        // Use cached balance
+        query.timestamp = { $gte: listAccountsSnapshot.createdAt };
       }
     }
-    return Array.from(uniqueAccounts);
+
+    const createdAt = new Date(); // take date earlier
+    const results = (await transactionModel.collection.distinct("accounts", query, {
+      session: options.session,
+    })) as string[];
+
+    let uniqueAccounts = fromDistinctToAccounts(results);
+
+    if (listAccountsSnapshot) {
+      uniqueAccounts = Array.from(new Set([...uniqueAccounts, ...listAccountsSnapshot.meta.accounts]));
+    }
+
+    if (uniqueAccounts.length === 0) return uniqueAccounts;
+
+    if (this.balanceSnapshotSec) {
+      // It's the first (ever?) snapshot for this listAccounts. We just need to save whatever we've just aggregated
+      // so that the very next listAccounts query would use cached snapshot.
+      if (!listAccountsSnapshot) {
+        await snapshotListAccounts({
+          book: this.name,
+          accounts: uniqueAccounts,
+          createdAt,
+          expireInSec: this.expireBalanceSnapshotSec,
+        });
+      } else {
+        // There is a snapshot already. But let's check if it's too old.
+        const tooOld = Date.now() > listAccountsSnapshot.createdAt.getTime() + this.balanceSnapshotSec * 1000;
+        // If it's too old we would need to cache another snapshot.
+        if (tooOld) {
+          delete query.timestamp;
+
+          // Important! We are going to recalculate the entire listAccounts from the day one.
+          // Since this operation can take seconds (if you have millions of documents)
+          // we better run this query IN THE BACKGROUND.
+          // If this exact balance query would be executed multiple times at the same second we might end up with
+          // multiple snapshots in the database. Which is fine. The chance of this happening is low.
+          // Our main goal here is not to delay this .listAccounts() method call. The tradeoff is that
+          // database will use 100% CPU for few (milli)seconds, which is fine. It's all fine (C)
+          transactionModel.collection
+            .distinct("accounts", query, {
+              session: options.session,
+            })
+            .then((results) => {
+              return snapshotListAccounts({
```

Comment:
hmm. why is the return here? iIs it here because then the .catch() can handle it?

Reply:
yes, correct.
I'll remove the `return` keyword and use the implicit return of arrow functions.